Changed jsonp to loadJsonp

ggetz opened the pull request:
Closes #2579

Created a new file `loadJsonp` and deprecated `jsonp`. Renamed occurrences of jsonp to loadJsonp. Updated CHANGED.md.

There was no Spec for jsonp, but I can add one for loadJsonp if needed.

pjcozzi commented:
> There was no Spec for jsonp, but I can add one for loadJsonp if needed.

Yes, please.

pjcozzi commented:
JsHint is failing: https://travis-ci.org/AnalyticalGraphicsInc/cesium/builds/85558063

pjcozzi commented:
You may be able to base your tests off of this: https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Specs/Core/loadJsonSpec.js

pjcozzi commented:
Looks good otherwise!

ggetz commented:
Added Test. I based it off the tests for `loadloadJsonSpec`, and removed what was not applicable, for example, when an error occurs, the error does not specify the error code or status.

pjcozzi commented on Specs/Core/loadJsonpSpec.js:
```diff
@@ -0,0 +1,40 @@
+/*global defineSuite*/
```
@mramato can you review these tests?

pjcozzi replied:
And merge if they are OK.

mramato commented on Source/Core/loadJsonp.js:
```diff
@@ -0,0 +1,118 @@
+
+            try {
+                delete window[functionName];
+            } catch (e) {
```
Not that you wrote this code, but delete will never raise an exception so we can remove the try/catch block.

shunter replied:
> delete will never raise an exception

Not strictly true.

```javascript
(function() { 
  "use strict"; 
  Object.defineProperty(window, 'foo', { configurable: false }); 
  delete window.foo; 
})();
```

Admittedly, this code was originally put here for IE8 compatibility reasons, I believe.

mramato replied:
Good point, (though it would be an extremely odd circumstance of events to get there).  I'm fine with leaving the code in just to be safe side if you want.

shunter replied:
I searched around to refresh my memory and yes, `delete window.foo` always throws on old IE.  http://stackoverflow.com/questions/1073414/deleting-a-window-property-in-ie  

So I'd suggest keeping it, since it doesn't seem too burdensome to do so.

mramato commented on Specs/Core/loadJsonpSpec.js:
```diff
@@ -0,0 +1,40 @@
+            expect(data).toEqual({
+                good : 'data'
+            });
+        }).otherwise(function(error) {
```
Since you aren't using the `otherwise` block, you can get rid of it.  Also, return the promise at the end of this function.  All of our asynchronous unit tests need to return the promise they create.

mramato commented on Specs/Core/loadJsonpSpec.js:
```diff
@@ -0,0 +1,40 @@
+
+    it('returns a promise that rejects when the request errors', function() {
+        var testUrl = 'http://example.invalid/testuri';
+        var promise = loadJsonp(testUrl).then(function(data){
```
You don't need to have an empty `then` block, you can just go right to `otherwise`.

mramato commented on Specs/Core/loadJsonpSpec.js:
```diff
@@ -0,0 +1,40 @@
+            expect(error).toBeDefined();
+        });
+
+        expect(promise).toBeDefined();
```
This will always be true, do you know why?

ggetz replied:
> This will always be true, do you know why?
Because if the request does not error the error block will not run?

mramato replied:
Not quite, it's because you already called `then` or `otherwise` on the result returned by `loadJsonp`, so if `loadJsonp` returned undefined an exception would have been generated already and execution would have bailed before getting to this line.

ggetz replied:
Ok, I understand, so like you said below, just return the call to `loadJsonp`.

mramato replied:
Yes.

mramato commented on Specs/Core/loadJsonpSpec.js:
```diff
@@ -0,0 +1,40 @@
+
+        });
+
+        expect(promise).toBeDefined();
```
This will also always be true, just return the `loadJsonp` line directly.

mramato commented:
I would also add a test that flexes each of the options objects and makes sure they work as expected.  Other than that and my couple of comments, this looks good.

mramato commented:
Thanks @ggetz

ggetz commented:
@mramato is there a way to simulate loading a url and getting a response with data? The methods used in `loadJsonSpec` doesn't work here, I believe.

ggetz commented:
Updated the unit tests.

mramato commented on Specs/Core/loadJsonpSpec.js:
```diff
@@ -0,0 +1,74 @@
+        loadJsonp(testUrl, options);
+    });
+
+
```
Sorry to nitpick, but we try to not leave extra whitespace/blank lines around.

mramato commented on Specs/Core/loadJsonpSpec.js:
```diff
@@ -0,0 +1,74 @@
+            proxy: new DefaultProxy(testProxy)
+        };
+        spyOn(loadJsonp, 'loadAndExecuteScript').and.callFake(function(url, functionName, deferred) {
+            expect(url).toStartWith(testProxy + '?' + testUrl);
```
I would replace `testProxy + '?' + testUrl` with `options.proxy.getURL(testUrl)`.  This ensures that you are testing that it's using the proxy object you pass into it and if `DefaultProxy` ever changes formats, this test will still pass and be valid.

mramato commented:
Just that one last comment.

ggetz commented:
@mramato Thanks, made the change.

mramato commented:
Thanks again @ggetz!
